core/src/main/java/com/yahoo/oak/NovaMemoryManager.java

```diff
@@ -0,0 +1,550 @@
+            return HEADER.lockWrite(version, blockID, offset, getMetadataAddress(), tap);
+        }
+
+        /**
+         * Releases a write lock
+         *
+         * @return {@code TRUE} if the write lock was released successfully
+         * {@code FALSE} if the value is marked as deleted
+         * {@code RETRY} if the value was moved, or the version of the off-heap value does not match {@code version}.
+         */
+        public ValueUtils.ValueResult unlockWrite() {
+            assert version != ReferenceCodecSyncRecycle.INVALID_VERSION;
+            return HEADER.unlockWrite(version, getMetadataAddress(),  tap);
+        }
+
+        /**
+         * Marks the associated off-heap cut as deleted only if the version of that value matches {@code version}.
+         *
+         * @return {@code TRUE} if the value was marked successfully
+         * {@code FALSE} if the value is already marked as deleted
+         * {@code RETRY} if the value was moved, or the version of the off-heap value does not match {@code version}.
+         */
+        public ValueUtils.ValueResult logicalDelete() {
+            assert version != ReferenceCodecSyncRecycle.INVALID_VERSION;
+            return HEADER.logicalDelete(version, getMetadataAddress());
+        }
+
+        /**
+         * Is the associated off-heap cut marked as logically deleted
+         *
+         * @return {@code TRUE} if the value is marked
+         * {@code FALSE} if the value is not marked
+         * {@code RETRY} if the value was moved, or the version of the off-heap value does not match {@code version}.
+         */
+        public ValueUtils.ValueResult isDeleted() {
+            assert version != ReferenceCodecSyncRecycle.INVALID_VERSION;
+            return HEADER.isLogicallyDeleted(version, getMetadataAddress());
+        }
+
+        /**
+         * Marks the header of the associated off-heap cut as moved, just write (without CAS)
+         * The write lock must be held (asserted inside the header)
+         */
+        public void markAsMoved() {
+            assert associated;
+            HEADER.markAsMoved(getMetadataAddress());
+            throw new IllegalAccessError();
+        }
+
+        /**
+         * Marks the header of the associated off-heap cut as deleted, just write (without CAS)
+         * The write lock must be held (asserted inside the header).
+         * It is similar to logicalDelete() but used when locking and marking don't happen in one CAS
+         */
+        public void markAsDeleted() {
+            assert version != ReferenceCodecSyncRecycle.INVALID_VERSION;
+            HEADER.logicalDelete(version, getMetadataAddress());
+        }
+
+    }
+    @Override
+    public <K> int compareKeyAndSerializedKey(K key, OakScopedReadBuffer serializedKey, OakComparator<K> cmp) {
+        if (((ScopedReadBuffer) serializedKey).s.lockRead() != ValueResult.TRUE) {
+            throw new DeletedMemoryAccessException();
```

Comment:
It might be already locked. Why do you assume it is deleted?
The same issue is below and in other MMs.

Reply:
In Nova MM the "lock" is not indeed a lock, and the only scenario that we can not "lock" is when the object is deleted.
hmmm any other names for the exception are welcomed!

Reply:
Then why aren't we using `isDeleted()` directly?
Does Nova not support mutual exclusion?